src/main/java/com/sourcegraph/cody/PromptPanel.kt

```diff
@@ -73,87 +84,222 @@ class PromptPanel(private val chatSession: ChatSession) : JLayeredPane() {
+      contextFilesSelectorModel = newModel
 
-  private fun getTextAndReset(): String {
-    val text = textArea.text
-    promptMessageHistory.messageSent(text)
-    textArea.text = ""
-    return text
+      if (newUserContextFiles.isNotEmpty()) {
+        contextFilesSelector.selectedIndex = 0
+      } else {
+        contextFilesSelector.selectedIndex = -1
+      }
+      refreshViewLayout()
+    }
   }
 
   companion object {
     private const val CHAT_MESSAGE_HISTORY_CAPACITY = 100
-    private val JUST_ENTER = KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), null)
+    private val KEY_ENTER = KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), null)
+    private val KEY_UP = KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_UP, 0), null)
+    private val KEY_DOWN = KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_DOWN, 0), null)
+    private val KEY_TAB = KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_TAB, 0), null)
+
+    val ENTER = CustomShortcutSet(KEY_ENTER)
+    val UP = CustomShortcutSet(KEY_UP)
+    val DOWN = CustomShortcutSet(KEY_DOWN)
+    val TAB = CustomShortcutSet(KEY_TAB)
+  }
+}
+
+data class DisplayedContextFile(val contextFile: ContextFile) {
+  override fun toString(): String {
+    return displayPath(contextFile)
+  }
+}
+
+data class AtExpression(
+    val startIndex: Int,
+    val endIndex: Int,
+    val rawValue: String,
+    val value: String
+)
+
+val atExpressionPattern = """(@(?:\\\s|[^\s])+)(?:\s|$)""".toRegex()
+
+fun findAtExpressions(text: String): List<AtExpression> {
+  val matches = atExpressionPattern.findAll(text)
+  val expressions = ArrayList<AtExpression>()
+  for (match in matches) {
+    val subMatch = match.groups.get(1)
+    if (subMatch != null) {
+      val value = subMatch.value.substring(1).replace("\\ ", " ")
+      expressions.add(
+          AtExpression(subMatch.range.first, subMatch.range.last + 1, subMatch.value, value))
+    }
+  }
+  return expressions
+}
+
+fun findContextFiles(contextFiles: List<ContextFile>, text: String): List<ContextFile> {
+  val atExpressions = findAtExpressions(text)
+  return contextFiles.filter { f -> atExpressions.any { it.value == displayPath(f) } }
+}
+
+// TODO(beyang): temporary displayPath implementation, should be updated to mirror what the VS Code
+// plugin does
+fun displayPath(contextFile: ContextFile): String {
+  // if the path contains more than three components, display the last three
+  val path = contextFile.uri.path
 
-    val UP = KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_UP, 0), null)
-    val DOWN = KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_DOWN, 0), null)
-    val DEFAULT_SUBMIT_ACTION_SHORTCUT = CustomShortcutSet(JUST_ENTER)
-    val POP_UPPER_MESSAGE_ACTION_SHORTCUT = CustomShortcutSet(UP)
-    val POP_LOWER_MESSAGE_ACTION_SHORTCUT = CustomShortcutSet(DOWN)
+  // split path on separator (OS agnostic)
+  val pathComponents = path.split(File.separator)
```

Comment:
It's better to avoid string operations on path.
More idiomatic way to do it could. be:
```kotlin
Paths.get("..").resolve(path.subpath(path.getNameCount() - 3, path.getNameCount()))
```

Reply:
Not sure I understand the recommendation here—converting `uri.path` to a `Path` object seems inadvisable, since the former uses posix style path separators, whereas `Path` seems like it would be OS-specific. This does flag an issue where I was using `File.separator` to split the path string, where  I should just be using `/` (and then replacing with `File.separator` in the return value).

Reply:
Ah, I haven't noticed that we start with uri and not path, and so we have are quarantined to have well defined schema we want to preserve.
You can ignore my comment then.